src/openms/source/ANALYSIS/OPENSWATH/DIAHelper.cpp

```diff
@@ -66,6 +66,151 @@ namespace OpenMS
+      typedef std::vector<double>::const_iterator itType;
+
+      itType mz_arr_end = spectrum->getMZArray()->data.end();
+      itType int_it = spectrum->getIntensityArray()->data.begin();
+      itType im_it = spectrum->getDriftTimeArray()->data.begin();
+
+      // this assumes that the spectra are sorted!
+      itType mz_it = std::lower_bound(spectrum->getMZArray()->data.begin(),
+          spectrum->getMZArray()->data.end(), mz_start);
+      itType mz_it_end = std::lower_bound(mz_it, mz_arr_end, mz_end);
+
+      // also advance intensity and ion mobility iterator now
+      std::iterator_traits< itType >::difference_type iterator_pos = std::distance((itType)spectrum->getMZArray()->data.begin(), mz_it);
+      std::advance(int_it, iterator_pos);
+      std::advance(im_it, iterator_pos);
+
+      // Iterate from mz start to end, only storing ion mobility values that are in the range
+      for (; mz_it != mz_it_end; ++mz_it, ++int_it, ++im_it)
+      {
+        if ( *im_it >= drift_start && *im_it <= drift_end)
+        {
+          intensity += (*int_it);
+          im += (*int_it) * (*im_it);
+        }
+      }
+
+      if (intensity > 0.)
+      {
+        im /= intensity;
+      }
+      else
+      {
+        im = -1;
+        intensity = 0;
+      }
+
+    }
+
+    bool integrateWindow(const OpenSwath::SpectrumPtr spectrum,
+                         double mz_start,
+                         double mz_end,
+                         double & mz,
+                         double & intensity,
+                         bool centroided)
+    {
+      OPENMS_PRECONDITION(spectrum->getMZArray()->data.size() == spectrum->getIntensityArray()->data.size(), "MZ and Intensity array need to have the same length.");
+      OPENMS_PRECONDITION(std::adjacent_find(spectrum->getMZArray()->data.begin(),
+              spectrum->getMZArray()->data.end(), std::greater<double>()) == spectrum->getMZArray()->data.end(),
+              "Precondition violated: m/z vector needs to be sorted!" )
+
+      mz = 0;
+      intensity = 0;
+      if (!centroided)
+      {
+        // get the weighted average for noncentroided data.
+        // TODO this is not optimal if there are two peaks in this window (e.g. if the window is too large)
+        typedef std::vector<double>::const_iterator itType;
+
+        itType mz_arr_end = spectrum->getMZArray()->data.end();
+        itType int_it = spectrum->getIntensityArray()->data.begin();
+
+        // this assumes that the spectra are sorted!
+        itType mz_it = std::lower_bound(spectrum->getMZArray()->data.begin(),
+          spectrum->getMZArray()->data.end(), mz_start);
+        itType mz_it_end = std::lower_bound(mz_it, mz_arr_end, mz_end);
+
+        // also advance intensity iterator now
+        std::iterator_traits< itType >::difference_type iterator_pos = std::distance((itType)spectrum->getMZArray()->data.begin(), mz_it);
```

Comment:
should we use auto instead of std::iterator_traits< itType >::difference_type here?